### ISSUE TITLE
systemd in containers not inheriting larger RLIMIT_NOFILE soft limit

### ISSUE THREAD

**Issue by anitazha** [systemd/systemd#19645]
**Is your feature request related to a problem? Please describe.**
If I do something like `sudo systemd-nspawn --rlimit=RLIMIT_NOFILE=1048576 --rlimit=RLIMIT_MEMLOCK=524288 -bi image.raw` with the default system.conf, `ulimit -a` in the container will not reflect the limits set on the command line. This applies to all containers running systemd as init, nspawn is just an example.

**Describe the solution you'd like**
I have 2 ideas that I was thinking about implementing but I'm not sure which is better (or maybe they're both terrible lol):
1. if systemd detects it's in a container it won't use any of it's internal defaults and always use what its parent set
2. if the DefaultLimit*= settings are explicitly set to empty in system.conf, then systemd won't use any internal defaults and will use what the parent/system sets

Both end goals are the same: to be able to set limits with `setrlimit()` during container set up and have the container inherit them. But (2) would also affect bare metal.

**Describe alternatives you've considered**
As part of container set up we could configure the DefaultLimit*= settings in system.conf to be the appropriate limits for each container. However we would prefer to ship system.conf as part of our base image and having to modify the container file system to change these limits seems not ideal. It's also unexpected that the limits aren't respected at all when system.conf is empty.

**The systemd version you checked that didn't have the feature you are asking for**
systemd 248+, current main branch f96bc66901ecbcf7ecb280d9442fd0cc92e67186

**Comment by poettering** [comment]
So the current code in pid1 actually leaves all rlimits untouched, except:

1. RLIMIT_NOFILE is increased to the maximum the kernel supports. It's left untouched if that's already the fact (or the limit is set higher already, which makes little sense, but you can still set)
2. RLIMIT_MEMLOCK is bumped to 1/8th of the available memory, except if already higher in which case nothing is done
3. RLIMIT_CORE is set to infinity


So the two limits you mention are among these three. What's your intention with setting that though? i.e. it appears you want to set them artificially low? what's the usecase for that?

I mean, usually people want to set them higher, not lower...

**Comment by anitazha** [comment]
In the case of RLIMIT_NOFILE we are trying to set it higher but the soft limit is not being updated properly. For example `sudo systemd-nspawn --rlimit=RLIMIT_NOFILE=1048576:1048576 -bi image.raw` inside nspawn `prlimit` shows
```
NOFILE     max number of open files             1024   1048576 files
```
RLIMIT_MEMLOCK does appear to be updating properly. If it's only RLIMIT_NOFILE that it affects then that's an easier issue to tackle. I can do a PR for this and add tests to make sure the others can be bumped too.

**Comment by poettering** [comment]
Oh, PID 1 explicitly sets the soft limit to 1024 because Linux sucks, basically.

The thing is that lots of code uses select() for watching for fd events, but that fails silently in bad ways if you try to watch an fd above 1K (aka FD_SETSIZE). Thus for compat with this code PID 1 will always set the soft limit to 1K at max, by default, to not break select() accidentally.

The expectation is that any program where it is sufficiently clear that select() is never called (neither by itself, nor by any libs it links to) and more than a bunch fds shell be something you want to deal with, expicitly opt into fds above 1024 by bumping their RLIMIT_NOFILE softlimit to the hardlimit early on. 

systemd's various tools do the same, we even have a helper call for that: `rlimit_nofile_bump()`.

You can also do the bumping inside the service unit files, but given that the select() usage is done in program code I'd also do the bumping in program code.

Long story short: bumping the soft limit for this for unprepared code is simply not safe. And if program code is not using select() then it should explicitly opt into higher fds via the equivalent of `rlimit_nofile_bump()`.

**Comment by anitazha** [comment]
Thanks for clearing that up I've opted for a man page update to make these defaults more apparent. But otherwise the code makes sense to me.

**Comment by poettering** [comment]
BTW, based on this discussion here I figured bloging about this made sense, and so I did:

http://0pointer.net/blog/file-descriptor-limits.html
